Learning path(fast): Add adaptive serving guide and codelab. Fixes #765, Fixes #665. (#857)
GoogleChrome/web.dev: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md

---
layout: codelab
title: Adapt video to image serving based on network quality
authors:
  - mihajlija
description: |
  Learn how to use the Network Information API to adapt
  your content based on the network quality.
glitch: adaptive-serving-netinfo-starter
related_post: adaptive-serving-based-on-network-quality
---

In this codelab, you’ll learn how to adapt your content based on the network
quality. This page's background video will load only when users are on a fast
network. On slower networks, an image will load instead.

The
[Network Information API](https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation)
enables you to access information about the user's connection quality. You will
use its `effectiveType` property to decide when to serve a video and when to
serve an image. `effectiveType` can be `'slow-2g'`, `'2g'`, `'3g'`, or `'4g'`.

## Step 1: Check connection type

> The background video is currently specified on line 22 of `index.html` and then

Reviewer: Also: since there's a longer explanation of this later on, maybe have a reference here, like, "The video loading code is described in more detail in [section]"

lazy-loaded in `script.js`.

To load it conditionally, first check if connection type detection is supported.
In `script.js` add an if statement that tests whether the `navigator.connection`

> object exists and whether it has the `effectiveType` property.

Reviewer: (Recommend putting `if` and `else` in code font.)

Then add an if statement to check the `effectiveType` of the network. To load
the video only on the fastest connections, it’s enough to check the connection
for `4g` and in any other case load the image.

```js
if (navigator.connection && !!navigator.connection.effectiveType) {
  if (navigator.connection.effectiveType === '4g') {
    // video loading code
  } else {
    // image loading code
  }
}
```

> Wrap the existing video loading code in an else statement, so that video will

Reviewer: ```suggestion
Wrap the existing video loading code in an `else` statement, so that video will
```

still load in browsers that don’t support the Network Information API.

```js
if (navigator.connection && !!navigator.connection.effectiveType) {
  if (navigator.connection.effectiveType === '4g') {
    // video loading code
  } else {
    // image loading code
  }
} else {
  const video = document.getElementById('coverVideo');
  const videoSource = video.getAttribute('data-src');
  video.setAttribute('src', videoSource);

  video.setAttribute('style', 'height: 100%; width: 100%; display:inline');
}
```

## Step 2: Load video

> If the `effectiveType` is `'4g'`, you can use the video loading code from the

Reviewer: ```suggestion
If the `effectiveType` is `'4g'`, use the video loading code from the
```

beginning of the codelab.

```js
if (navigator.connection.effectiveType === '4g') {
  const video = document.getElementById('coverVideo');
  const videoSource = video.getAttribute('data-src');
  video.setAttribute('src', videoSource);
  video.setAttribute('style', 'height: 100%; width: 100%; display:inline');
} else {
  // image loading code
}
```

> Here’s how the video loading code works: the video source is specified inside

Reviewer: ```suggestion
Here’s how the video loading code works: the video source is specified using the
```

Reviewer: Alternately, maybe explain a little more here. Like:

.. code works: the `<video>` tag doesn't download or display anything at first because its `src` attribute is not set. The video URL to load is specified using the `data-src` attribute.

> [data-src attribute](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes)

Reviewer: Don't need two links to this doc here. Recommend maybe cutting this one and only including the following one. Also: `data-src` should be in code font.

> in `index.html` and then lazy-loaded in `script.js`.

Reviewer: ```suggestion
on the `<video>` tag.
```

```html
<video id="coverVideo" autoplay loop muted data-src="https://cdn.glitch.com/b6491350-b058-4eb6-aa6c-55c93122073e%2FMatrix%2C%20Console%2C%20Hacking%2C%20Code.mp4?1551464245607"></video>
```

Because the video source is specified in the `data-src` attribute, it isn’t
displayed or downloaded initially.
[Data attributes](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes)

> allow you to store extra information on standard HTML elements. They can be

Reviewer: ```suggestion
allow you to store extra information on standard HTML elements. A data element can be
```

named anything, as long as it starts with "data-".

To actually display the video on the page, the location from `data-src` needs to

> be set as the `src` attribute of the video element. This is done in `script.js`.

Reviewer: "the location from ..." maybe rephrase as "you need to copy the value from `data-src` to the video element's `src` attribute."

I think . you can leave off the "This is done in script.js."

video.setAttribute('style', 'height: 100%; width: 100%; display:inline');
```

## Step 3: Load image

To conditionally load an image on slower networks, use the same strategy as for
the video.

Add an image element to `index.html` (right after the video element), and use
the `data-src` attribute instead of the `src` attribute.

```html
<img id="coverImage" data-src="https://cdn.glitch.com/36529535-5976-40f8-979b-40c898b86bd0%2F36529535-5976-40f8-979b-40c898b86bd0_1_Sn80dgiwpMjBVrqjfiDbnA.jpg?1553003835358" />
```

In `script.js`, add code to set the image’s `src` attribute depending on the
`effectiveType` of the network.

```js
if (navigator.connection.effectiveType === '4g') {
  const video = document.getElementById('coverVideo');
  const videoSource = video.getAttribute('data-src');
  video.setAttribute('src', videoSource);

  video.setAttribute('style', 'height: 100%; width: 100%; display:inline');
} else {
  const image = document.getElementById('coverImage');
  const imageSource = image.getAttribute('data-src');
  image.setAttribute('src', imageSource);

  image.setAttribute('style', 'height: 100%; width: 100%; display:inline');
}
```

## Try it out

To test it yourself:

> 1. Mouse over the editor, press the **App** button, then press the **Show**

Reviewer: "press the **App** button  ... " can be shortened to "click **App**, then click **Show** to preview the app."

   button to preview the app.
2. Press `CTRL + SHIFT + J` or `CMD + OPTION + J` (Mac), to open DevTools.

> 3. Click on the **Network tab**.

Reviewer: ```suggestion
3. Click the **Network** tab.
```

> 4. Select **Fast 3G** from throttling presets.

Reviewer: ```suggestion
4. Click the **Throttling** dropdown, which is set to **No throttling** by default. Select  **Fast 3G**.
```

<img class="w-screenshot" src="./devtools_network_throttling.png" alt='DevTools Network tab with Fast 3G throttling option highlighted'>

> Now reload the page with Fast 3G still enabled and the app will load an image in the background instead of the video:

Reviewer: ```suggestion
Now reload the page with Fast 3G still enabled. The app loads an image in the background instead of the video:
```

<img class="w-screenshot" src="./netinfo_app_image.png" alt='Matrix-like image background with "NETWORK INFORMATION" text overlay'>

## Extra Credit: Respond to changes

Remember how this API has an `onchange`

> [event listener](http://localhost:8080/fast/adaptive-serving-based-on-network-quality#how-it-works)?

Reviewer: Bad URL. (I think just cut http://localhost:8080 and make it an absolute path?)

> You can use it for many things: dynamically adapt content such as video quality, restart deferred data transfers when a change to a high-bandwidth network type is detected, or notify users when the network quality changes.

Reviewer: ```suggestion
You can use it for many things: dynamically adapting content such as video quality; restarting deferred data transfers when a change to a high-bandwidth network type is detected; or notifying users when the network quality changes.
```

> Here’s a simple example of how this listener works, add it to `script.js`. This

Reviewer: ```suggestion
Here’s a simple example of how to use this listener. Add it to `script.js`. This
```

code will call the `displayNetworkInfo` function whenever the network
information changes.

```js
navigator.connection.addEventListener('change', displayNetworkInfo);
```

There’s already an empty `<h2>` element on the `index.html` page. Now define the
`displayNetworkInfo` function so that it displays the network information in the
`<h2>` element and invoke the function.

```js
function displayNetworkInfo() {
  document.getElementById('connection').innerHTML = navigator.connection.effectiveType;
}

displayNetworkInfo();
```

Here’s the final state of the [app on Glitch](https://glitch.com/~adaptive-serving-netinfo).

<img class="w-screenshot" src="./netinfo_app_video_background.png" alt='Matrix-like video background with "NETWORK INFORMATION 4g" text overlay'>

To test it again:

> 1. Mouse over the editor, press the **App** button, then press the **Show**

Reviewer: ```suggestion
1. Mouse over the editor, click **App** button, then click **Show**
```

>    button to preview the app.

Reviewer: ```suggestion
   to preview the app.
```

2. Press `CTRL + SHIFT + J` or `CMD + OPTION + J` (Mac), to open DevTools.

> 3. Click on the **Network tab**.

Reviewer: ```suggestion
3. Click the **Network** tab.
```

> 4. Select **Fast 3G** from throttling presets.

Reviewer: ```suggestion
4. Click the **Throttling** dropdown, which is set to **No throttling** by default. Select  **Fast 3G**.
```

> 5. **Reload** the page.

Reviewer: ```suggestion
5. Reload the page.
```
